Using BufferedFsInputStream to wrap FSInputStream for FSDataInputStream

n3nash opened the pull request:
This PR is not ready for any review.

n3nash commented on hoodie-client/src/test/java/com/uber/hoodie/common/HoodieTestDataGenerator.java:
```diff
@@ -69,7 +70,7 @@ public HoodieTestDataGenerator(String[] partitionPaths) {
 
   public HoodieTestDataGenerator() {
-    this(new String[] {"2016/03/15", "2015/03/16", "2015/03/17"});
+    this(new String[]{"2016/03/15", "2015/03/16", "2015/03/17"});
```
Not sure why this is happening, my reformat does this.

n3nash commented on hoodie-common/src/main/java/com/uber/hoodie/common/table/log/HoodieLogFileReader.java:
```diff
@@ -237,6 +241,10 @@ private boolean isBlockCorrupt(int blocksize) throws IOException {
       inputStream.seek(currentPos + blocksize);
     } catch (EOFException e) {
       // this is corrupt
+      // This seek is required because contract of seek() is different for naked DFSInputStream vs BufferedFSInputStream
```
This was one of the most tedious things to track down, phew

n3nash commented on hoodie-client/src/test/java/com/uber/hoodie/io/TestHoodieCommitArchiveLog.java:
```diff
@@ -42,23 +42,56 @@
   private Configuration hadoopConf;
+  //NOTE : Be careful in using DFS (FileSystem.class) vs LocalFs(RawLocalFileSystem.class)
+  //The implementation and gurantees of many API's differ, for example check rename(src,dst)
+  // We need to use DFS here instead of LocalFs since the FsDataInputStream.getWrappedStream() returns a FsDataInputStream
```
Another extremely tedious thing to track down, wish LocalFS and DFS had same contracts :(

n3nash commented:
@vinothchandar @bvaradar I'm keeping the BufferedInputStream changes local to the LogReader for now. We can open another issue whether we want to make this change in HoodieWrapperFileSystem. 
Also, I'm open to suggestions for the DEFAULT_BUFFER_SIZE. I don't feel like making it configurable to user since it's something more internal. I'm going to test with 8 MB in the production system tomorrow.

n3nash commented:
Results from testing in production

**8 MB block size** 

**18/04/05 21:17:34** INFO log.HoodieLogFileReader: Log HoodieLogFile {.....} has a corrupted block at 14
**18/04/05 21:18:04** INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {....} starts at 8240687
18/04/05 21:18:04 INFO log.HoodieLogFileReader: Log HoodieLogFile {.....} has a corrupted block at 14
18/04/05 21:18:34 INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {.....} starts at 8240687

**64 MB block size**

**18/04/05 21:41:29** INFO log.HoodieLogFileReader: Log HoodieLogFile {....} has a corrupted block at 14
**18/04/05 21:41:56** INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {....} starts at 8240687
18/04/05 21:41:57 INFO log.HoodieLogFileReader: Log HoodieLogFile {....} has a corrupted block at 14
18/04/05 21:42:31 INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {....} starts at 8240687

**256 MB block size**

**18/04/05 21:49:20** INFO log.HoodieLogFileReader: Log HoodieLogFile {....} has a corrupted block at 14
**18/04/05 21:49:48** INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {....} starts at 8240687
18/04/05 21:49:48 INFO log.HoodieLogFileReader: Log HoodieLogFile {....} has a corrupted block at 14
18/04/05 21:50:17 INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {....} starts at 8240687

**256 MB block size old PR** 

**18/04/05 22:36:41** INFO log.HoodieLogFileReader: Log HoodieLogFile {....} has a corrupted block at 14
**18/04/05 22:37:09** INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {....} starts at 8240687
18/04/05 22:37:10 INFO log.HoodieLogFileReader: Log HoodieLogFile {....} has a corrupted block at 14
18/04/05 22:37:34 INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {....} starts at 8240687

Summary for corrupt block of around 8.5MB

| Block_Size | Approx Time Taken (secs) |
|-------------|--------------------------|
| 8 MB          | 90  |
| 64 MB | 25 |
| 256 MB | 28 |

vinothchandar commented:
So something like 64MB should work? its still kind of weird though 64MB seeking takes 20 sec. Is it due to some sys call overhead

n3nash commented:
Yeah, 20 secs is way too large, I'm going to delve into what's going on soon. BTW, 64 MB works in this example since the corrupted bytes are > 8 MB and < 64 MB. At any point when the corrupted bytes are > BLOCK_SIZE, we will see a much higher amount of time taken to find the corrupt block.

n3nash commented:
Tried a few things like use System.copy(..) instead of inputStream.readFully(..) twice but doesn't reduce the time by a lot. Essentially, each byte results in a read of (6 bytes + 4 bytes) for each MAGIC header. So, a 8.5 MB corrupt block results in 10*8000000 bytes corresponding all operations for each byte (compare MAGIC headers) etc. To avoid large times for corrupt blocks the best buffer size is to set to BLOCK_SIZE which is 256 MB. This will add a 256MB memory overhead. The build is failing with OOM, looking into why that's happening on jenkins not on my local.

n3nash commented:
@vinothchandar Please take a pass at this now. Updated time taken : 

| Block_Size | Approx Time Taken (secs) |
|-------------|--------------------------|
| 16 MB          | 1  |
| 64 MB | 1 |
| 256 MB | 1 | 

16 MB block

18/04/15 22:54:33 INFO log.HoodieLogFileReader: Log HoodieLogFile {....log.1} has a corrupted block at 14
18/04/15 22:54:33 INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {....log.1} starts at 8240687
18/04/15 22:54:34 INFO log.HoodieLogFileReader: Log HoodieLogFile {...log.1} has a corrupted block at 14
18/04/15 22:54:35 INFO log.HoodieLogFileReader: Next available block in HoodieLogFile {..log.1} starts at 8240687

The extra time spent was in the fact that for each byte scan the code was throwing an Exception. Throwing and catching exceptions is an expensive operation. Some more info here : http://java-performance.info/throwing-an-exception-in-java-is-very-slow/. I've refactored the code to avoid throwing exceptions and instead return boolean.

vinothchandar commented on hoodie-cli/src/main/java/com/uber/hoodie/cli/commands/HoodieLogFileCommand.java:
```diff
@@ -172,7 +175,8 @@ public String showLogFileRecords(@CliOption(key = {
           Boolean.valueOf(HoodieCompactionConfig.DEFAULT_COMPACTION_LAZY_BLOCK_READ_ENABLED),
-          Boolean.valueOf(HoodieCompactionConfig.DEFAULT_COMPACTION_REVERSE_LOG_READ_ENABLED));
+          Boolean.valueOf(HoodieCompactionConfig.DEFAULT_COMPACTION_REVERSE_LOG_READ_ENABLED),
+          Integer.valueOf(HoodieMemoryConfig.DEFAULT_MAX_HDFS_STREAM_BUFFER_SIZE));
```
rename to : `..MAX_DFS_STREAM_BUFFER_SIZE`

n3nash replied:
done

vinothchandar commented on hoodie-cli/src/main/java/com/uber/hoodie/cli/commands/HoodieLogFileCommand.java:
```diff
@@ -92,6 +92,9 @@ public String showLogFileCommits(
           try {
             instantTime = n.getLogBlockHeader().get(HeaderMetadataType.INSTANT_TIME);
+            if (instantTime == null) {
+              throw new Exception("Invalid value " + instantTime + " for instant time ");
```
change to  : "Invalid instant time" + null

n3nash replied:
done

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -45,6 +45,9 @@
   // Property to set the max memory for compaction
   public static final String MAX_MEMORY_FOR_COMPACTION_PROP = "hoodie.memory.compaction.max.size";
+  // Property to set the max memory for hdfs inputstream buffer size
+  public static final String MAX_HDFS_STREAM_BUFFER_SIZE_PROP = "hoodie.memory.hdfs.buffer.max.size";
```
rename here as well

n3nash replied:
done

n3nash commented on hoodie-hadoop-mr/src/main/java/com/uber/hoodie/hadoop/realtime/HoodieRealtimeRecordReader.java:
```diff
@@ -76,9 +76,14 @@
+  // Property to set the max memory for hdfs inputstream buffer size
+  public static final String MAX_HDFS_STREAM_BUFFER_SIZE_PROP = "hoodie.memory.hdfs.buffer.max.size";
+  // Setting this to lower value of 1 MB since no control over how many RecordReaders will be started in a mapper
+  public static final int DEFAULT_MAX_HDFS_STREAM_BUFFER_SIZE = 1 * 1024 * 1024; // 1 MB
```
@vinothchandar Kept this lower since don't know memory settings for mapper. WDYT ?

n3nash commented:
@vinothchandar made renames and left one comment

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -45,6 +45,9 @@
   // Property to set the max memory for compaction
   public static final String MAX_MEMORY_FOR_COMPACTION_PROP = "hoodie.memory.compaction.max.size";
+  // Property to set the max memory for hdfs inputstream buffer size
+  public static final String MAX_DFS_STREAM_BUFFER_SIZE_PROP = "hoodie.memory.hdfs.buffer.max.size";
```
hoodie.memory.dfs.buffer.max.size?

n3nash replied:
ahh, apologies for the oversight, will rename correctly

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieMemoryConfig.java:
```diff
@@ -86,6 +89,12 @@ public Builder withMaxMemoryFractionPerCompaction(long maxMemoryFractionPerCompa
       return this;
     }
 
+    public Builder withMaxStreamBufferSize(int maxStreamBufferSize) {
```
rename: withMaxDFSStreamBufferSize(..)

n3nash replied:
done

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieWriteConfig.java:
```diff
@@ -369,6 +369,12 @@ public Long getMaxMemoryPerCompaction() {
             props.getProperty(HoodieMemoryConfig.MAX_MEMORY_FOR_COMPACTION_PROP));
   }
 
+  public int getMaxBufferSize() {
```
make consistent with `withMaxDFSStreamBufferSize` above?

n3nash replied:
done
